pyk/src/pyk/kore/syntax.py

```diff
@@ -2170,6 +2170,7 @@ def axioms(self) -> tuple[Axiom, ...]:
 class Definition(Kore, WithAttrs, Iterable[Module]):
     modules: tuple[Module, ...]
     attrs: tuple[App, ...]
+    ordinals: dict[int, Axiom]
```

Comment:
This makes the class no longer immutable and hashable. Instead, consider making this a `cached_property`:

```python3
@cached_property
def ordinals(self) -> FrozenDict[int, Axiom]:
    ...  # calculate ordinals
```

Then `get_axiom_by_ordinal` becomes unnecessary.

Reply:
Hey Tamás, I'm not sure if you want to replace `get_axiom_by_ordinal` or `compute_ordinals` here. 
One is computing and modifying the definition,  other is just returning the element in the cached dict. Can you elaborate your suggestion?  

The `compute_ordinals` is one step on `preprocess` LLVM function for context.

Reply:
`ordinals: dict[int, Axiom]` should not be an attribute of the class in this form as it is
* mutable
* not hashable

Where does this value come from btw?

Reply:
Got this part! Is `FrozenDict` better in this sense, right?

This value is computed from `compute_ordinals` by counting the Axioms. We not only have this same [logic in the LLVM backend definition](https://github.com/runtimeverification/llvm-backend/blob/0b0bdaeaf5acfed9034c40bec68b285dc5fac4dc/lib/ast/definition.cpp#L183-L198) but we also have tools to compute and use it from the `definition.kore`, see: [here](https://github.com/runtimeverification/llvm-backend/tree/master/tools/llvm-kompile-compute-ordinal) and [here](https://github.com/runtimeverification/llvm-backend/tree/master/tools/llvm-kompile-compute-loc).

Reply:
And the ordinal of the `n`-th `Axiom` is simply `n`, right? Because in that case, storing the `dict` isn't necessary:

```python3
@cached_property
def axioms(self) -> tuple[Axiom, ...]:
    return tuple(sent for module in self for sent in module if isinstance(sent, Axiom))

def get_axiom_by_ordinal(self, ordinal: int) -> Axiom:
    return self.axioms[ordinal]
```

Reply:
Perfect. The problem here is that I really need a new definition containing the ordinal attribute in its axioms to work with. This was previously only an LLVM Backend attribute, but we have to have it here as well, as we're using PyK to parse the LLVM Proof Hints. Eventually, we'll upstream this attribute to the future Python `ModuleToKore` class, so this last party should be temporary.

Reply:
Constructing a new definition is not a problem. And since it preserves the order of axioms, you can compute the `axiom` list from either of the definitions.

Reply:
Right, I'm just saying that using this approach, we'll pass through all the sentences of all modules twice when needed: one for computing the `axioms` and another for creating the new definition.

Reply:
If you're okay with that, that's fine with me; I understand this modification is softer than adding a new field to `Definition`. I just want to highlight this rework when using these functions together.

Reply:
Is iterating over sentences twice a performance bottleneck? If so, you can return the `dict` along with the `Definition` from `compute_ordinals`. You can introduce a custom class to encapsulate the two values.

If there's no known performance issue then you can just iterate twice and go with the simpler design.